File: backend/src/main/java/com/sweetpos/entity/Sale.java
package com.sweetpos.entity;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "sales")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Sale extends BaseEntity {

    @Column(nullable = false, unique = true, length = 30)
    private String invoiceNumber;

    @Column(length = 50)
    private String terminalId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "cashier_id", nullable = false)
    private User cashier;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id")
    private Customer customer;

    @Column(nullable = false, precision = 14, scale = 2)
    private BigDecimal subtotal;

    @Column(nullable = false, precision = 12, scale = 2)
    @Builder.Default
    private BigDecimal discountAmount = BigDecimal.ZERO;

    @Column(length = 20)
    private String discountType;

    @Column(precision = 8, scale = 2)
    @Builder.Default
    private BigDecimal discountValue = BigDecimal.ZERO;

    @Column(nullable = false, precision = 12, scale = 2)
    @Builder.Default
    private BigDecimal taxAmount = BigDecimal.ZERO;

    @Column(nullable = false, precision = 14, scale = 2)
    private BigDecimal totalAmount;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private PaymentMethod paymentMethod;

    @Column(precision = 14, scale = 2)
    private BigDecimal cashReceived;

    @Column(precision = 14, scale = 2)
    private BigDecimal changeAmount;

    @Column(nullable = false)
    @Builder.Default
    private int loyaltyEarned = 0;

    @Column(nullable = false)
    @Builder.Default
    private int loyaltyRedeemed = 0;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private SaleStatus status = SaleStatus.COMPLETED;

    @Column(length = 500)
    private String notes;

    @OneToMany(mappedBy = "sale", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<SaleItem> items = new ArrayList<>();

    // Used for sync
    @Column(length = 100)
    private String localId;

    public enum PaymentMethod {
        CASH, CARD, QR, MIXED
    }

    public enum SaleStatus {
        COMPLETED, CANCELLED, REFUNDED, PENDING_SYNC
    }
}

File: flutter_app/lib/models/sale.dart
class Sale {
  final String? id;
  final String? localId;
  final String? invoiceNumber;
  final String? terminalId;
  final String cashierId;
  final String? customerId;
  final double subtotal;
  final double discountAmount;
  final String? discountType;
  final double discountValue;
  final double taxAmount;
  final double totalAmount;
  final String paymentMethod;
  final double? cashReceived;
  final double? changeAmount;
  final int loyaltyEarned;
  final int loyaltyRedeemed;
  final String status;
  final String? notes;
  final List<SaleItem> items;
  final DateTime createdAt;
  final bool synced;

  Sale({
    this.id,
    this.localId,
    this.invoiceNumber,
    this.terminalId,
    required this.cashierId,
    this.customerId,
    required this.subtotal,
    this.discountAmount = 0,
    this.discountType,
    this.discountValue = 0,
    this.taxAmount = 0,
    required this.totalAmount,
    required this.paymentMethod,
    this.cashReceived,
    this.changeAmount,
    this.loyaltyEarned = 0,
    this.loyaltyRedeemed = 0,
    this.status = 'COMPLETED',
    this.notes,
    required this.items,
    required this.createdAt,
    this.synced = false,
  });

  Map<String, dynamic> toCreateJson() => {
        'localId': localId,
        'customerId': customerId,
        'terminalId': terminalId ?? 'FLUTTER-01',
        'items': items.map((i) => i.toJson()).toList(),
        'discountType': discountType,
        'discountValue': discountValue,
        'discountAmount': discountAmount,
        'paymentMethod': paymentMethod,
        'cashReceived': cashReceived,
        'loyaltyRedeemed': loyaltyRedeemed,
        'notes': notes,
      };
}

class SaleItem {
  final String productId;
  final String? productName;
  final int quantity;
  final double unitPrice;
  final double discountAmount;
  final double taxAmount;
  final double totalPrice;

  SaleItem({
    required this.productId,
    this.productName,
    required this.quantity,
    required this.unitPrice,
    this.discountAmount = 0,
    this.taxAmount = 0,
    required this.totalPrice,
  });

  Map<String, dynamic> toJson() => {
        'productId': productId,
        'quantity': quantity,
        'discountAmount': discountAmount,
      };
}
